### PR TITLE
Proposal for switching desktop-client to vite

### PR DESCRIPTION
**Pull request by twk3**
Proposal for vite switch for https://github.com/actualbudget/actual/issues/656

## Why we need to switch from CRA/craco

They appear to be deprecated, and not getting their dependencies updated. That's a problem because it makes it a lot more difficult to keep actual's dependencies up-to-date. It's currently difficult to update to the latest typescript as an example.

## What we want

- something low maintenance (no ejecting react-scripts, avoid manual config of all the skeleton dependencies)
- something well maintained, and receiving dependency updates
- minimal deviation from existing work to begin with (either drops in nicely, or has a piece-by-piece migration path)

## Pros of vite

- One of the most popular cra replacements (along with nextjs, and remix), and as a result actively maintained and updated.
- Was very easy to swap in and get the dev instance working
- quick dev watches as changed files get dynamically reloaded
- Supports swc so we aren't changing quite as much
- vitest is based on jest, and is close to a drop in replacement
- Final bundle size is similar, (a bit smaller) than our existing

## Cons of vite

- Still changing, major version changes mean we will be making changes while it stabilizes
- Swaps our bundler/minimizer (to rollup), and css handing (a con vs being able to stay in the webpack ecosystem as an iteration)
- the swc plugin doesn't do minify (there is a swc-only plugin avail that does, but the one I'm using here is the official plugin, in order to maintain good compat with rollup/vite going forward)
- The use of esbuild and rollup fragments our monorepo (which is still using webpack)
- Their doesn't appear to be any good chunking plugins avail in as our filesize grows
- Doesn't really offer us an iterative migration path
-Will need to build our own custom transform for the github action size compare (as it only works for webpack). This is outputing the stats, but we don't have a consumer.
- There are likely to be a ton of little edge cases broken that the tests aren't catching

## Speed and size

- desktop-client bundles are a MB or two smaller with vite. 
- Production bundle speeds are smiliar, currently vite is a  second slower on my machine, but some extra times is caused by the shim plugins I have put in place to match more of our old behaviour. Without the env shim vite production builds are a second faster. Once this is merged and we know we aren't moving back to craco, we can remove the shims and adjust the code.
- Dev server is way faster. Always less than half a second with vite. With webpack, first serve is about 3 seconds, first reload is about 13 (I don't know why), and then further reloads can be as low as half a second. 
  * Timed used desktop client only, for webpack using timing info from stats.json
  * I did also do some test runs with filesystem cache turned on for webpack, but that didn't change the timings on my machine (I have lots of memory on my dev machine, so the in memory caching was probably working fine)
  * Other than that first reload our current webpack setup actually isn't slow. But vite feels instant, which makes sense, it's making the browser do all the work in this mode. The moment you save your file the browser basically starts loading the change.

### Major changes

- the dev server is sending modules direct to the browser for loading, no compiling involved in dev mode (very fast, but very different)
- rollup instead of webpack, esbuild instead of swc for minification and css (swc is still being used for transformations)
- vitest instead of jest
-  an incredible amount of dependency changes

### CODE REVIEW

**Comment by joel-jeremy** [comment]
This is exciting! Took a stab before at implementing Vite bit but encountered issues with the inter-ui fonts as well as resolution errors.

**Comment by MatissJanis** [comment]
👋 Could you list out the pros/cons of the switch to vite? Do we have any performance, bundle size, DX benefits here? What cons would the switch have?

And don't get me wrong: I'm not _against_ this move. I just want to get a better sense of the potential drawbacks of this approach.

A while ago we used to have loads of manually defined webpack files. They were a nightmare to maintain.. and upgrading webpack versions was super labour-intensive. So I made the decision to switch to react-scripts. Whilst it was not _ideal_ - it did give us back a lot of dev capacity to focus on the product instead of maintaining build scripts.

No matter the future tool we choose: I would prefer that it is low-maintenance. So we get to spend our limited time working on the product, not the tooling around it.

**Comment by twk3** [comment]
Fixed the netlify build so we can see it actually working. There is probably some better way that would allow us to keep using process.env, but currently process.env is always an empty object in a prod build, rather than being turned static)

@MatissJanis I updated the description with some goals/pros cons that I've thought of.

Of note this was just me trying a vite migration to see how it went, but it did actually go better than I was expecting, and might be worth consideration. (I'm happy to attempt a similar migration with another provider if someone wants to suggest some). My main goal is exploring ways to unlock dependency updates. react-scripts appears dead, and was getting in the way of trying a typescript upgrade. My first approach was to eject, but I did read the PRs introducing CRA back in, and craco, and agree with the desire to reduce the maintenance burden.  

As a result my hope was to actually find a react-scripts replacement that was still using webpack. With swc in there, I don't see a reason to change if possible. But I wasn't able to find a popular one that was still being maintained. (things like snowpack are suggesting vite). I was actually planning to see what remix, (and still calling out to webpack from it), would look like. But a recent comment in https://github.com/actualbudget/actual/issues/656 caused me to make an attempt at vite first. 

I have to say, if you don't look at the yarn.lock, it actually dropped in pretty well. The big config file is mostly me trying to match the old build output and old config file. (so that this just drops into actual-server perfectly for example without any changes needed).

There are lots of downsides, when you do look at the lockfile, you can see the expanse of the behind the scenes changes this swap does. If we merged this, we would be spending the next month  fixing edge cases we broke. And then ongoing, vite is still maturing, and releasing major versions somewhat frequently. We don't get those nice dependency updates along for the ride unless we also update vite. (this MR already needs a new vite update just to fix the fonts).

That being said, I do think this shows that vite is a viable option. And we will need something here in the next year when we want to update the dependencies. Let me know if there are others you want me to POC.

**Comment by glowtape** [comment]
Small comment, since I recently switched to Vite. Running Vite as dev server, it'll use esbuild, which in turn uses the React SWC plugin. In this case, minify is kinda irrelevant. If you build and bundle the project, that's all rollup instead, which doesn't use that plugin.

**Comment by joel-jeremy** [comment]
Is there a Vite plugin to allow us to continue using process.env? I am most curious about the difference in build times and dev server reload time between the current and Vite so it would be great if we can measure that.

**Comment by twk3** [comment]
> Small comment, since I recently switched to Vite. Running Vite as dev server, it'll use esbuild, which in turn uses the React SWC plugin. In this case, minify is kinda irrelevant. If you build and bundle the project, that's all rollup instead, which doesn't use that plugin.

@glowtape in this MR swc is also used for transform even in build mode. This is triggered by passing a swc specific plugin the vite-react-swc plugin. No clue if there is a benefit yet. I was trying to keep the output similar to what we already have, which is using swc.

**Comment by twk3** [comment]
> Is there a Vite plugin to allow us to continue using process.env? I am most curious about the difference in build times and dev server reload time between the current and Vite so it would be great if we can measure that.

@joel-jeremy I went ahead and embedded one that fixes it so now process.env is working here.  Of note both here, and previously, this was NOT the node process.env. (I saw a lot of stackoverflow suggestions to just dump node process env into the defines). Both CRA and vite are careful to clear out process.env. CRA just reused it for it's own env handling as well, whereas vite uses import.meta.env.   


The inject plugin gets it back in there, but I was previously running into issues with the fact that vite was coming through during build and purposely replacing any instance of process.env with an empty object. https://github.com/vitejs/vite/blob/main/packages/vite/src/node/plugins/define.ts#L20 

It's working now. Was all about timing things around when vite was messing with process.env so it didn't wipe out the shim.

Build times and size were similar to what we had before. Slightly longer on the build times (9 seconds vs 8 seconds on my machine, due to the inject plugin) And slightly smaller on the output sizes. 

I haven't done any experiments with dev server times, which is where we SHOULD be getting some benefit.

**Comment by MatissJanis** [comment]
Thanks for your write-up @twk3 ! 

Personally: as long as we can get feature parity with our current build set-up I would be in favour of switching over to Vite. It would unblock us from upgrading some outdated devtools. And I also have a hunch that it might solve https://github.com/actualbudget/actual/issues/1766 .

**Comment by twk3** [comment]
@MatissJanis I moved the renames over into a MR here that can be reviewed and merged if wanted: https://github.com/actualbudget/actual/pull/2101  turns on the eslint rule to enforce it the (t|j)sx filename.

**Comment by twk3** [comment]
Fonts are worked around now, and this has been rebased to include the renames that went in seperately. Now it's easier to review the actual changes here.

I haven't divined into why vite or its css is like this, but it's not resolving css urls relative to the node_module the css file is in. Using import directly of a scss file in the module works, but doing it via `use` does not. 

We have some options to work around, what I've done here is just set the font path variable to go to the proper location. But we could also switch to an `import` from our index.tsx and drop the fonts.scss.  (we would need to update inter-ui first for the import approach, as the current version would import all the fonts, vs the new version would just do the ones we use, similar to the current 'use' approach).

**Comment by twk3** [comment]
After looking around and not really finding an good alternative that didn't involve a third party monitoring service, I ported and published a github action that gives us a sizecompare for rollup. https://github.com/marketplace/actions/rollup-comparison (just the webpack one, but modified to map the stats file from rollup visualizer)

I also removed the manual chunks. As the default is already patching the number that webpack had (with a bit smaller size), and maintaining a manual list seemed like a pain. I had only added them because rollup was complaining about the chunk sizes, so I've bumped up the warning threshold.

Remaining items that I know of is to look into the victory warnings, and see if I can see a dev environment load improvement.

**Comment by twk3** [comment]
Tracked down the victory warnings. Looks like it's an incompatibility between babel output and rollup's PURE annotation implementation. I logged an issue with rollup here: https://github.com/rollup/rollup/issues/5324

The impact (other than the warnings), is that tree shaking isn't working to remove extra code in the victory components. While nice to fix, the fact that our bundle sizes are already smaller with rollup probably makes this a non-issue for us. I'm going to take a look at filtering out these warnings for now to reduce some noise.

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx [hunk 1]
```diff
@@ -367,7 +367,7 @@ const TransactionHeader = memo(
         />
         {showBalance && <Cell value="Balance" width={88} textAlign="right" />}
         {showCleared && <Field width={23} truncate={false} />}
-        <Cell value="" width={5 + scrollWidth ?? 0} />
+        <Cell value="" width={scrollWidth ? 5 + scrollWidth : 0} />
```
Comment (twk3) [inline packages/desktop-client/src/components/transactions/TransactionsTable.jsx:370]: This is the only visual change, (causing a visual test failure). The esbuild was just complaining that the existing change was a bug. (would never return 0). Should move this fix to a different PR to completely eliminate visual changes from this PR!

Comment (joel-jeremy) [reply packages/desktop-client/src/components/transactions/TransactionsTable.jsx:370]: Would `<Cell value="" width={5 + (scrollWidth ?? 0)} />` fix the warning and the visual regression?

Comment (twk3) [reply packages/desktop-client/src/components/transactions/TransactionsTable.jsx]: @joel-jeremy indeed you are correct, we don't want this to be 0, we want it to be at least 5, and plus the scrollWidth if available. Updated.

**Comment by twk3** [comment]
Renamed to a propoal, and this is now ready for review.

Today I did some speed comparisons, which I added a section to the description for. This is much faster in dev mode when  you are changing a desktop-client file. But I don't feel like we were really that bad before.

I also noticed today that there were some instructions where we were using the HTTPS env variable and getting auto-generated self-signed certificates on craco/CRA. I went ahead and added a plugin that gives us that same behaviour here as well.

**Comment by joel-jeremy** [comment]
Please add the release notes so that the CI passes. :)

**Comment by twk3** [comment]
Release note added!

Also want to point out, that while it looks like there is a massive drop in dependencies in yarn.lock due to this change (and there IS a massive swap of react-script dependencies to vite ones), the reduction in dependencies is more related to the fact that no one has run `yarn dedupe` in a while.

File: packages/desktop-client/vite.config.mts
Comment (joel-jeremy) [inline packages/desktop-client/vite.config.mts:1]: Question: Is there a reason why mts was chosen as extension here instead of ts/js?

**Comment by joel-jeremy** [comment]
Trying this out locally, building and starting the app is working great! But for some reason, debugging (e.g. breakpoints) from VS Code doesn't seem to be working for me.

**Comment by MatissJanis** [comment]
Looks like the electron build is no longer working with this on OSX. `yarn build:desktop`

**Comment by MatissJanis** [comment]
> Looks like the electron build is no longer working with this on OSX. `yarn build:desktop`

Actually.. I recommend holding off on doing deeper investigations into this. The current websocket implementation is very flaky. I'm moving it back to IPC. Once that is done - the issue we are seeing here in vite might go away as well.

File: packages/desktop-client/vite.config.mts
Comment (twk3) [reply packages/desktop-client/vite.config.mts:1]: @joel-jeremy Vite throws a warning on startup otherwise. With the alternative being to put `type: module` into the package.json for desktop-client. This seemed like the smaller change https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated

**Comment by MatissJanis** [comment]
#2190 is now merged. If you don't mind updating this PR - I can run more tests on Electron and other platforms to make sure things still work with vite.

**Comment by twk3** [comment]
> Trying this out locally, building and starting the app is working great! But for some reason, debugging (e.g. breakpoints) from VS Code doesn't seem to be working for me. Is there a step required to configure debugging with vite?

@joel-jeremy it's working the same for me in this MR as previous, at least for desktop-client breakpoints. What were you running into? 

My launch.json is (chrome default with just the port and the webroot updated):

```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "type": "chrome",
            "request": "launch",
            "name": "Launch Chrome against localhost",
            "url": "http://localhost:3001",
            "webRoot": "${workspaceFolder}/packages/desktop-client/src"
        }
    ]
}
```

And I also typically launch the dev server in debug via the start option in the top package.json file in vscode. Then after it's up, I go to the debug section and tell it to launch chrome against localhost.

What's the process that is giving you trouble?

**Comment by joel-jeremy** [comment]
@twk3 Turns out I needed to update the webRoot. It's working now! My original configuration was:
```json
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "chrome",
      "request": "launch",
      "name": "Launch Chrome against localhost",
      "url": "http://localhost:3001",
      "webRoot": "${workspaceFolder}"
    }
  ]
}
```

**Comment by MatissJanis** [comment]
Sadly electron app is still not working on this branch 😢 

Try running it: `yarn start:desktop`. For me it's just showing the loading spinner forever (osx). Is it working for you?

**Comment by twk3** [comment]
@MatissJanis it initially loads for me, but I get the spinner as soon as I try do do anything beyond the budget list. Like I can start an import, but the moment we get to the spinner it just all stops. (same with loading a budget or creating a test budget.) Deleting budgets from the budget list is working. Will need to look further.

**Comment by twk3** [comment]
Bizarrely it's hanging in the backend. in sqlite.openDatabase.

**Comment by twk3** [comment]
It's some sort of issue with one of the dependency changes unrelated to vite. Running yarn dedupe on the master branch repros the issue. Will need to dive in to figure out which dependency it is.

**Comment by twk3** [comment]
The issue was deduping node-abi. Electron and better-sqlite have an issue with this as described here: https://github.com/WiseLibs/better-sqlite3/issues/1044  . Updating to electron 27 looks like it fixes this. But for the sake of this PR I've just reduped the node-abi in the yarn.lock to be the same as master.

@MatissJanis I believe the electron issue should now be fixed.

**Comment by joel-jeremy** [comment]
Would it be ideal to upgrade to electron 27 now (in a separate PR of course)? Or would that require a lot of other changes?

**Comment by MatissJanis** [comment]
FYI: I'm doing the Electron upgrade as part of https://github.com/actualbudget/actual/pull/2182 . Lucky coincidence I guess :) 

Will try to re-review this PR again later today.

**Review by MatissJanis — DISMISSED** [review]
I am not seeing any more issues. This new build process is WAY faster. In both local-dev, but also CI.

LGTM on my end, but I'd prefer to have 1-2 more maintainers review this since it's such a core part of the build process.

**Comment by MatissJanis** [comment]
Merged the electron PR. Apologies for creating merge conflicts here @twk3

**Comment by twk3** [comment]
Added one more commit. Rollup released a new version that fixed the victory warnings https://github.com/rollup/rollup/issues/5324 so I was able to drop the warning suppression we had for that from the vite config

**Review by joel-jeremy — APPROVED** [review]
LGTM! Awesome work!

**Review by MatissJanis — APPROVED** [review]
